File: src/model/runtime-model.ts
import { createProviderRegistry, type LanguageModel } from "ai";
import { createAnthropic } from "@ai-sdk/anthropic";
import { createOpenAI } from "@ai-sdk/openai";

import type { ModelClient } from "../agent/model-client.js";
import { DemoModel } from "../demo/demo-model.js";
import {
  readRuntimeModelConfig,
  type DevRuntimeModelConfig,
  type ModelRuntimeEnvironment,
} from "./runtime-config.js";
import {
  type GenerateStructuredAction,
  VercelAiModelClient,
} from "./vercel-ai-model-client.js";

type CreateRuntimeModelOptions = {
  env?: ModelRuntimeEnvironment;
  resolveDevLanguageModel?: (config: DevRuntimeModelConfig) => LanguageModel;
  generateStructuredAction?: GenerateStructuredAction;
};

export function createRuntimeModel(
  options: CreateRuntimeModelOptions = {},
): ModelClient {
  const env = options.env ?? process.env;
  const config = readRuntimeModelConfig(env);
  if (config.mode === "demo") {
    return new DemoModel();
  }

  return new VercelAiModelClient({
    model: (options.resolveDevLanguageModel ?? resolveDevLanguageModel)(config),
    modelId: config.modelId,
    generateStructuredAction: options.generateStructuredAction,
  });
}

export function resolveDevLanguageModel(config: DevRuntimeModelConfig): LanguageModel {
  if (config.provider === "openai") {
    const registry = createProviderRegistry({
      openai: createOpenAI({
        apiKey: config.apiKey,
        baseURL: config.baseURL,
      }),
    });

    return registry.languageModel(config.modelId as `openai:${string}`);
  }

  const registry = createProviderRegistry({
    anthropic: createAnthropic({
      apiKey: config.apiKey,
      baseURL: config.baseURL,
    }),
  });

  return registry.languageModel(config.modelId as `anthropic:${string}`);
}

File: agent_language_lab/model/runtime_model.py
from __future__ import annotations

from typing import Mapping

from agent_language_lab.agent.model_client import ModelClient
from agent_language_lab.demo.demo_model import DemoModel
from agent_language_lab.model.runtime_config import (
    ModelRuntimeEnvironment,
    read_runtime_model_config,
)


def create_runtime_model(
    env: ModelRuntimeEnvironment | None = None,
) -> ModelClient:
    config = read_runtime_model_config(env)
    if config.mode == "demo":
        return DemoModel()

    raise NotImplementedError(
        "Python runtime dev mode is planned for Phase 2. "
        "Current Python implementation supports AGENT_MODEL_MODE=demo only."
    )
